gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/webauthn/WebAuthnResponseEndpoint.java

```diff
@@ -196,4 +271,56 @@ private void updateCredential(AuthenticationContext authenticationContext, Strin
                         error -> handler.handle(Future.failedFuture(error))
                 );
     }
+
+    private void updateSessionAuthAndChallengeStatus(RoutingContext ctx) {
+        ctx.session().put(ConstantKeys.STRONG_AUTH_COMPLETED_KEY, true);
+        ctx.session().put(ConstantKeys.MFA_CHALLENGE_COMPLETED_KEY, true);
+    }
+
+    private void updateSessionLoginCompletedStatus(RoutingContext ctx, User user, String credentialId){
+        // save the user and login completed status into the context
+        ctx.getDelegate().setUser(user);
+        ctx.session().put(ConstantKeys.PASSWORDLESS_AUTH_COMPLETED_KEY, true);
+        ctx.session().put(ConstantKeys.WEBAUTHN_CREDENTIAL_ID_CONTEXT_KEY, credentialId);
+        ctx.session().put(USER_LOGIN_COMPLETED_KEY, true);
+    }
+
+
+
+    private void redirectToAuthorize(RoutingContext ctx){
+        final MultiMap queryParams = RequestUtils.getCleanedQueryParams(ctx.request());
+        final String returnURL = getReturnUrl(ctx, queryParams);
+        ctx.response().putHeader(HttpHeaders.LOCATION, returnURL).end();
+    }
+
+    private boolean isEnrollingFido2Factor(RoutingContext ctx) {
+        final String factorId = ctx.session().get(ENROLLED_FACTOR_ID_KEY);
+        if(factorId == null) {
+            return false;
+        }
+
+        return FactorType.FIDO2.getType().equals(factorManager.getFactor(factorId).getFactorType().getType());
+    }
+
+    private Completable saveFidoFactor(io.gravitee.am.model.User user, String credentialId, String fidoFactorId) {
+        final Factor factor = factorManager.getFactor(fidoFactorId);
+        final EnrolledFactor enrolledFactor = createEnrolledFactor(factor.getId(), credentialId);
+
+        try {
+            return userService.addFactor(user.getId(), enrolledFactor, new DefaultUser(user)).ignoreElement();
+        } catch (Exception ex) {
+            return Completable.error(ex);
+        }
+    }
+
+    private EnrolledFactor createEnrolledFactor(String factorId, String credentialId){
```

Comment:
Seems to be a duplicate with this method 
https://github.com/gravitee-io/gravitee-access-management/pull/1869/files#diff-c7814c24f32d59c21c2966fbdbe6d0b21966b5c21c6e7b6c5cbf3e1ff5eae919R551

Reply:
Most of the duplicate methods now moved to FidoFactorUtils class.